### PR TITLE
Implementing NEON-ized Adler32 checksum

### PR DESCRIPTION
**Pull request by Adenilson**
The checksum is calculated in the uncompressed PNG data and can be made much faster by using SIMD. Tests in ARMv8 yielded an improvement of about 3x (e.g. walltime was 350ms x 125ms for a 4096x4096 bytes executed 30 times).

This yields an improvement in image decoding in Chromium around 18% (see https://bugs.chromium.org/p/chromium/issues/detail?id=688601).

### CODE REVIEW

**Comment by Adenilson** [comment]
I'm still unable to build it as adler32.c will fail to include the path that has the adler32_neon.h.

Any tips about what has to change in the configure script to fix that?

File: adler32.c [hunk 1]
```diff
@@ -7,6 +7,10 @@
 
 #include "zutil.h"
 
+#if (defined(__ARM_NEON__) || defined(__ARM_NEON))
+#include "adler32_neon.h"
```
Comment (Dead2) [inline adler32.c]: Put the function definition here, then there is no need for having a adler32_neon.h at all.
As I mentioned on IRC, this is the way it is done elsewhere in the code. That will also solve the compilation problem.

**Comment by Adenilson** [comment]
Making the suggested changes will fix the compilation error, but then later there will be a linker error.

For some reason the binaries for neonized Adler are not being compiled and added when 'ar' executes.

Manually changing to the folder (e.g. arch/aarch64) and running 'make' will build the binaries (e.g. adler32_neon.o|lo).

Any tips on what may be missing in the build script? I've run ./configure with --neon flag.

**Comment by Adenilson** [comment]
The CMake buildsystem (as expected) is easier to tweak.

Is there a special reason to use armv8a-simd instead of "-march=armv8-a -mfpu=neon"? The later is better supported by older compilers (e.g. gcc 5.4).

**Comment by Adenilson** [comment]
Never ever tried building for ARM using Windows before, it may be a caveat of the linaro compiler for MS Windows.

Using the simd flag will fail with gcc version 5.4.0 20160609 (Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4) *while* running in Linux@ARM.

For Chromium it doesn't matter as everything is built with GN/ninja (https://chromium.googlesource.com/chromium/src/tools/gn/) and they won't be using MS Windows to build for Android.

**Comment by Adenilson** [comment]
Should be good now (...famous last words...).

File: arch/aarch64/Makefile.in [hunk 2]
```diff
@@ -48,10 +54,13 @@ depend:
 
 # DO NOT DELETE THIS LINE -- make depend depends on it.
 
+adler32_neon.o: $(SRCDIR)/adler32_neon.h
+adler32_neon.lo: $(SRCDIR)/adler32_neon.h
```
Comment (mtl1979) [inline arch/aarch64/Makefile.in]: Line 58 should be moved above line 63.

File: arch/arm/Makefile.in [hunk 3]
```diff
@@ -48,10 +54,13 @@ depend:
 
 # DO NOT DELETE THIS LINE -- make depend depends on it.
 
+adler32_neon.o: $(SRCDIR)/adler32_neon.h
+adler32_neon.lo: $(SRCDIR)/adler32_neon.h
```
Comment (mtl1979) [inline arch/arm/Makefile.in]: Line 58 should be moved above line 63.

**Comment by Dead2** [comment]
Thanks!

**Pull request merged**